Extensions/separator.js

```diff
@@ -1,212 +1,570 @@
-//* VERSION 1.1.3 **//
-//* DESCRIPTION Where were we again? **//
-//* DEVELOPER STUDIOXENIX **//
-//* DETAILS A simple extension that puts a divider showing where you left off on your dashboard. **//
+//* VERSION 2.0.0 **//
+//* DESCRIPTION Marks where you left off on your dashboard **//
+//* DEVELOPER new-xkit **//
 //* FRAME false **//
+//* SLOW false **//
 //* BETA false **//
 
-XKit.extensions.separator = new Object({
-
-	running: false,
-
-	check_for: "",
-
-	preferences: {
-
-		"bold_line": {
-			text: "Use the thick line for separating",
-			default: false,
-			value: false
-		},
-
-		"jump_to": {
-			text: "Show the \"Jump to Separator\" button (only works when endless scrolling is off)",
-			default: false,
-			value: false
-		},
-
-	},
-
-	run: function() {
-		this.running = true;
-
-		if (XKit.interface.where().dashboard !== true) {return; }
-
-		XKit.tools.init_css("separator");
-
-		if (XKit.extensions.separator.preferences.bold_line.value === true) {
-
-			XKit.tools.add_css(" #xkit-post-separator { border-top: 3px solid rgba(255,255,255,0.33) !important; border-bottom: 0 !important; } ", "separator-thick");
+// # Motivation
+// Some of us like to catch up with all the posts that we’ve missed since we last checked our dashboard.
+// Placing a visual separator to show where we left off helps with that.
+//
+// # Approach
+// For each context (dashboard, specific blog, or specific tag) we track the newest post we’ve previously witnessed.
+// This is our “goal post” (above which we’ll place the separator).
+// When we return to the context, we track the set of posts that we’ve witnessed that’s newer than the goal post.
+// Once we reach the goal post, we place the separator.
+// We then reset the goal post (to the newest post in the contiguous interval of witnessed posts that’s touching the goal post).
+//
+// # Implementation-specific Caveats
+//   - We don’t actually track the set of discrete witnessed posts; we track a set of disjoint intervals representing contiguously witnessed posts.
+//     This approach should tolerate deleted posts.
+//     However, a witnessed post that’s subsequently been edited will still be treated as witnessed.
+//   - When endless scrolling is off, merging of seemingly disjoint intervals from adjacent pages relies on URL hints.
+//     If these hints are missing, we’ll fail to merge these intervals.
+//     Consequently, once the goal post resets, we’ll find it pointing to an older than expected post.
+//     The workaround is to manually trigger a goal post reset (with the separator controls).
+//     Hints might be missing if we manually browse to a page’s URL.
+//     They might also _go_ missing should Tumblr change their URL patterns.
+//   - When the same context is open in multiple tabs, behaviour may seem unpredictable.
+
+XKit.extensions.separator = (function() {
+	"use strict";
+
+	// Immutables
+	const DASHBOARD_PATH_REGEX = /^\/dashboard(?:\/(?:([2-9]|[1-9]\d+)\/(-?[1-9]\d*)\/?)?)?$/;
```

Comment:
is there a non-trivial functionality we gain by using [2-9]|[1-9]\d+? I understand that it's more correct, but \d+ seems like it would work just fine in the practical case.

also when do we ever see negative post numbers?

Reply:
¯\\\_(ツ)\_/¯

I figured it was a common enough construct for matching against positive integers, and have a preference towards failing early. And the pattern is unlikely to change with frequency for maintainability to be an issue. (I mean, I feel the escapes and nested optional groups are what make this regex a pain to read.) But I can change it to `\d+`.

Negative post numbers are for backwards paging.